File: src/launch/requestLaunchLambda.ts
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import { SendMessageCommand, SQSClient } from '@aws-sdk/client-sqs';
import process from 'process';
import pino from 'pino';

import LaunchRequest = Components.Schemas.LaunchRequest;
import LaunchResponse = Components.Schemas.LaunchResponse;

const sqsClient = new SQSClient({ region: 'eu-central-1' })
const logger = pino()

const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
    logger.info('Incoming launch request')
    const requestBody = event.body;
    if (!requestBody) {
        throw Error('Missing launch request')
    }
    const { rocketName, destination } = JSON.parse(requestBody) as LaunchRequest;

    await sqsClient.send(
        new SendMessageCommand({
            QueueUrl: process.env['LAUNCH_QUEUE_URL'],
            MessageBody: JSON.stringify({ rocketName: rocketName })
        })
    )
    const status = 'LAUNCHING'

    return {
        statusCode: 200,
        body: JSON.stringify({
            rocketName,
            destination,
            status
        } as LaunchResponse)
    }
}

module.exports = { handler }

File: src/launch/request_launch_lambda.py
import json
import logging
import os
import urllib.request

logger = logging.getLogger()
logger.setLevel(logging.INFO)


def handler(event, context):
    logger.info("Incoming launch request")

    request_body = event.get("body")
    if not request_body:
        raise ValueError("Missing launch request")

    body = json.loads(request_body)
    rocket_name = body["rocketName"]
    destination = body["destination"]

    mission_control_base_url = os.environ["MISSION_CONTROL_BASE_URL"]
    req = urllib.request.Request(
        f"{mission_control_base_url}/mission/{rocket_name}", method="GET"
    )
    with urllib.request.urlopen(req) as response:
        mission = json.loads(response.read().decode())

    status = "UNDERWAY" if mission["progress"] > 0 else "LAUNCHING"

    return {
        "statusCode": 200,
        "body": json.dumps(
            {
                "rocketName": rocket_name,
                "destination": destination,
                "status": status,
            }
        ),
    }
